src/mgr/ClusterState.h

```diff
@@ -27,6 +28,24 @@ class MMgrDigest;
 class MMonMgrReport;
 class MPGStats;
 
+class C_ClientInMgrMapCond : public C_SaferCond {
+  std::string_view name;
+  entity_addrvec_t client_addr;
+public:
+  C_ClientInMgrMapCond(std::string_view name_, entity_addrvec_t client_addr_) :
+    C_SaferCond("C_ClientInMgrMapCond"), name(name_), client_addr(client_addr_) {}
+
+  bool check_client_in_mgrmap(const MgrMap& mgrmap) {
+    auto itp = mgrmap.clients.equal_range(std::string(name));
```

Comment:
Since mgrmap.clients is a multimap that allows duplicate key-value pairs, I think we'll need to keep track of the count of identical key-value pairs on registration? Can the count of duplicate key-value pairs change when waiting for changes to MgrMap, e.g. a client with identical kv pair deregisters? Need to think about this.

It'd have been nice if the mgrmap.clients was a std::map<std::string,  std::vector<entity_addrvec_t>> or just a plain std::map<std::string, entity_addrvec_t> forcing the module name's client to have a unique name?

Reply:
> Since mgrmap.clients is a multimap that allows duplicate key-value pairs, I think we'll need to keep track of the count of identical key-value pairs on registration? Can the count of duplicate key-value pairs change when waiting for changes to MgrMap, e.g. a client with identical kv pair deregisters? Need to think about this.

You can have duplicate keys but not duplicate key-values (TMK).

> It'd have been nice if the mgrmap.clients was a std::map<std::string, std::vector<entity_addrvec_t>> or just a plain std::map<std::string, entity_addrvec_t> forcing a module's client to have a unique name?

At the time it was expected some modules would have multiple rados instances so a multimap made sense to me since it would not insert duplicate key-value pairs.

Reply:
> You can have duplicate keys but not duplicate key-values (TMK).
> 

Please see https://stackoverflow.com/a/5632419 . Multimaps can have duplicate key value pairs

> > It'd have been nice if the mgrmap.clients was a std::map<std::string, std::vector<entity_addrvec_t>> or just a plain std::map<std::string, entity_addrvec_t> forcing a module's client to have a unique name?
> 
> At the time it was expected some modules would have multiple rados instances so a multimap made sense to me since it would not insert duplicate key-value pairs.

Got it.

Reply:
Well, that makes sense in retrospect. There is no value comparator so how could it check for uniqueness?

Anyway, if you want to rework that as a `std::map<std::string, std::set<entity_addrvec_t>>`, I would support that change. The encoding/decoding would not need to change.

Reply:
> I think the goal should be a simple `std::map<std::string, entity_addrvec_t>`. When we were adding names to the mgrmap in #50065, it was with the expectation that the name would uniquely identify a given RADOS client (for observability, to be able to selectively blocklist in tests, etc). The modules with multiple RADOS instances could supply a tag and the actual name would be `<module name>.<tag>` or similar; for non-cooperating modules, the framework could generate a random string tag.

I like the above the idea. @batrick what do you think?

Reply:
Sounds reasonable to me but some mgr modules may need adapted to the new design.

Reply:
> > I think the goal should be a simple `std::map<std::string, entity_addrvec_t>`. When we were adding names to the mgrmap in #50065, it was with the expectation that the name would uniquely identify a given RADOS client (for observability, to be able to selectively blocklist in tests, etc). The modules with multiple RADOS instances could supply a tag and the actual name would be `<module name>.<tag>` or similar; for non-cooperating modules, the framework could generate a random string tag.
> 
> I like the above the idea. @batrick what do you think?

I spent more time thinking about this. I think changing `clients` to type  std::map<std::string, entity_addrvec_t> doesn't look feasible. We can't maintain backwards compatibility when decoding MgrMap from old mgr. See the case, https://github.com/ceph/ceph/pull/50065/commits/b545fb9f5660dca3af4dea195ea4555f09b3a6e8#diff-4939dba11a535c85c49f5ae03d0c217cf4c9497b1b33876327eada5a93775de6R489
Currently, when we have a new monitor and old manager, the names of modules are empty strings in the `mgr dump` output 
```
$ ceph mgr dump | jq '.active_clients'
[
  {
    "name": "",
    "addrvec": [
      {
        "type": "v2",
        "addr": "127.0.0.1:0",
        "nonce": 165306198
      }
    ]
  }
]
```
To continue handling this scenario,  I think  `std::map<std::string, std::set<entity_addrvec_t>> clients` in mon/MgrMap.h,  messages/MMgrBeacon.h and mgr/PyModuleRegistry.h makes better sense to me. @idryomov what do you think?

Reply:
I think this means that we are stuck with a multimap for practical purposes.  I mean, we could make the backwards compatibility `""` a special case and see into transforming other duplicates but it's really not worth the trouble IMO.

I commented just because I don't see much advantage in `std::map<std::string, std::set<entity_addrvec_t>>` over `std::multimap<std::string, entity_addrvec_t>`.  If `std::map<std::string, entity_addrvec_t>` isn't feasible, the current multimap is totally fine.